lib/integrations/samlidp/gcpworkforce.go

```diff
@@ -0,0 +1,215 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package samlidp
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"net/http"
+	"time"
+
+	"github.com/gravitational/trace"
+	"google.golang.org/api/googleapi"
+	iam "google.golang.org/api/iam/v1"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/integrations/samlidp/samlidpconfig"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// GCPWorkforceService defines workforce service configuration
+// parameters.
+type GCPWorkforceService struct {
+	// APIParams holds basic input params required to create GCP workforce
+	// pool and pool provider.
+	APIParams samlidpconfig.GCPWorkforceAPIParams
+	// HTTPClient is used to fetch metadata from the SAMLIdPMetadataURL
+	// endpoint.
+	HTTPClient *http.Client
+}
+
+// NewGCPWorkforceService creates a new GCPWorkforceService.
+func NewGCPWorkforceService(cfg GCPWorkforceService) (*GCPWorkforceService, error) {
```

Comment:
Receiving a type T and returning the same type T does not seem like the usual builder pattern we follow.
I would create a companion Config struct or just pass the `samlidpconfig.GCPWorkforceAPIParams` and the `http.Client`

Reply:
I do not think there's anything wrong with receiving a type T and returning same type T. Builder pattern is also used for input validation which is what this func does.

FYI, I only added this new func so to add `checkAndSetdefault` for input validation. 

I have updated comment which now says "// NewGCPWorkforceService creates a new GCPWorkforceService with input validation." in https://github.com/gravitational/teleport/pull/39820/commits/91ae579c03203bf0d1d878c3d13b4ac51cb4d2ea if that helps.